src/soter/boringssl/soter_kdf.c

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2019 Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "soter/soter_kdf.h"
+
+#include <limits.h>
+
+#include <openssl/evp.h>
+
+soter_status_t soter_pbkdf2_sha256(const uint8_t* passphrase,
+                                   size_t passphrase_length,
+                                   const uint8_t* salt,
+                                   size_t salt_length,
+                                   size_t iterations,
+                                   uint8_t* key,
+                                   size_t key_length)
+{
+    int res;
+
+    if (!passphrase) {
+        SOTER_CHECK_PARAM(passphrase_length == 0);
+    }
```

Comment:
why don't check `SOTER_CHECK_PARAM(passphrase_length <= INT_MAX);` here too as it did for openssl?

Reply:
That's because of a difference in argument types:

| Soter                                   | OpenSSL                             | BoringSSL                          |
| --------------------------------------- | ----------------------------------- | ---------------------------------- |
| `soter_status_t soter_pbkdf2_sha256(`   | `int PKCS5_PBKDF2_HMAC(`            | `int PKCS5_PBKDF2_HMAC(`           |
| `    const uint8_t* passphrase,`        | `    const char *         pass,`    | `    const char*    password,`     |
| `    size_t         passphrase_length,` | `    int                  passlen,` | `    size_t         password_len,` |
| `    const uint8_t* salt,`              | `    const unsigned char* salt,`    | `    const uint8_t* salt,`         |
| `    size_t         salt_length,`       | `    int                  saltlen,` | `    size_t         salt_len,`     |
| `    size_t         iterations,`        | `    int                  iter,`    | `    unsigned       iterations,`   |
|                                         | `    const EVP_MD*        digest,`  | `    const EVP_MD*  digest,`       |
| `    size_t         key_length`         | `    int                  keylen,`  | `    size_t         key_len,`      |
| `    uint8_t*       key,`               | `    unsigned char*       out`      | `    uint8_t*       out_key`       |
| `);`                                    | `);`                                | `);`                               |

Soter and BoringSSL use mostly the same types for arguments, in particular they both use `size_t` for sizes. We don't need many checks and casts here. However, where OpenSSL uses `int` we need to ensure that Soter argument will fit into that type.